abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1996 @@
+
+#### `wasi_snapshot_preview1.random_get`
+
+* params:
+  - `i32 (uint8_t *) buffer`
+  - `i32 (size_t) buffer_size`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+Generates `buffer_size` bytes of randomness.
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+- `INVAL` when requested `buffer_size` is too large.
+- `FAULT` when `buffer` and/or `buffer_size` point to invalid memory
+  address.
+
+
+## Environment variables
+
+> **Warning**
+> Environment variables should be configured for each plugin.
+> Exposing the host's environment variables is highly discouraged.
+
+
+### Functions exposed by the host
+
+#### `wasi_snapshot_preview1.environ_sizes_get`
+
+* params:
+  - `i32 (size_t *) return_num_elements`
+  - `i32 (size_t *) return_buffer_size`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+Returns `return_num_elements` of environment variables, and
+`return_buffer_size` sufficient to serialize them and decorators.
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+- `FAULT` when `return_num_elements` and/or `return_buffer_size`
+  point to invalid memory address.
+
+
+#### `wasi_snapshot_preview1.environ_get`
+
+* params:
+  - `i32 (uint8_t **) return_array`
+  - `i32 (uint8_t *) return_buffer`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+Returns serialized environment variables.
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+- `FAULT` when `return_array` and/or `return_buffer` point to
+  invalid memory address.
+
+
+## Buffers
+
+Access to buffers using functions in this section is restricted to
+specific callbacks:
+
+- `HTTP_REQUEST_BODY` can be read and modified in
+  [`proxy_on_request_body`] (or for as long as request processing
+  is paused from it).
+- `HTTP_RESPONSE_BODY` can be read and modified in
+  [`proxy_on_response_body`] (or for as long as response processing
+  is paused from it).
+- `DOWNSTREAM_DATA` can be read and modified in
+  [`proxy_on_downstream_data`] (or for as long as upstream processing
+  is paused from it).
+- `UPSTREAM_DATA` can be read and modified in
+  [`proxy_on_downstream_data`] (or for as long as downstream processing
+  is paused from it).
+- `HTTP_CALL_RESPONSE_BODY` can be read in
+  [`proxy_on_http_call_response`].
+- `GRPC_CALL_MESSAGE` can be read in [`proxy_on_grpc_receive`].
+- `VM_CONFIGURATION` can be read in [`proxy_on_vm_start`].
+- `PLUGIN_CONFIGURATION` can be read in [`proxy_on_configure`].
+- `FOREIGN_FUNCTION_ARGUMENTS` can be read in
+  [`proxy_on_foreign_function`].
+
+
+### Functions exposed by the host
+
+#### `proxy_set_buffer_bytes`
+
+* params:
+  - `i32 (`[`proxy_buffer_type_t`]`) buffer_id`
+  - `i32 (uint32_t) offset`
+  - `i32 (uint32_t) size`
+  - `i32 (const uint8_t *) value_data`
+  - `i32 (size_t) value_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Sets content of the buffer `buffer_id` to the provided value
+(`value_data`, `value_size`) replacing `size` bytes starting
+at `offset` in the existing buffer.
```

Comment:
I think we should document prepend/append/replace behavior based on the offset/size combination as in: https://github.com/envoyproxy/envoy/blob/a0620080acf46534675bc6484c8127ff1c2d0898/source/extensions/common/wasm/context.cc#L129-L149